libflux/Polymorphic_Labels.md

```diff
@@ -0,0 +1,103 @@
+
+
+```flux
+builtin fill : (<-tables: stream[A], ?column: string, ?value: B, ?usePrevious: bool) => stream[C]
+    where
+    A: Record,
+    C: Record
+```
+
+However as we can see from `fill`'s signature, the type system is wholly unaware of whether the incoming record (`A`) has the `column`. It also doesn't know what the shape of the returned record (`C`) is either. To fix this issue we need a way to connect the `column` value with the input (and output records). As a solution I propose we add "polymorphic record labels" which would let us define `fill` as follows.
+
+```flux
+builtin fill : (<-tables: stream[{ A with 'column: B }], ?column: 'column, ?value: B, ?usePrevious: bool) => stream[{ A with 'column: B }]
+    where A: Record
+```
+
+`'column` in this case indicates a polymorphic label which would allow `fill` to be called as normal, however when passing a "label" (a string) that is known at compile time it lets us enforce the typing of the input and output records.
+
+```flux
+
+// 'column is inferred to `a` which propagates to the input and output records, making them `{ A with a: B }`
+[{ a: 1 }] |> fill(column: "a", value: 0)
+
+// ERROR: the input record lacks the field `a`
+[{}] |> fill(column: "a", value: 0)
+```
+
+
+## Explanation
+
+Polymorphic labels adds the ability to define record types where the field names (labels) can be a variable instead of a literal name in a similar way that type variables can exist in a type.
+
+```
+{ 'col: string } vs { some_column: string }
+(x: A) => A vs (x: string) => string
+```
+
+The types that fit into a record field implement the new `Label` kind to ensure that unexpected types like `Int` are an error when used in a record field.
+
+To make the change as transparent as possible string literals now get the new `Label("literal")` type instead of `String` making `"abc": Literal("abc")`. While this means that two different string literals will have different types we must still treat `Label` types as strings in most cases. Consider
+
+```flux
+if b then "a" else "b"
+```
+
+The branches will have the `Label("a")` and `Label("b")` respectively, which we still need to unify successfully and the resulting type of the expression should still be a `string`. As long as we limit  the uses of label types to direct uses in calls (like in `fill` above). We may keep the same type checking behaviour by treating `Label` as a `String` types in every instance except when type checking function calls, in which case we specialize unification to a subsumption check where we allow some limited sub-typing.
+
+```flux
+// The original fill call is allowed such that `'column` is inferred to be `Label("a")`
+fill(column: "a", value: 0)
+
+// We should (probably) still allow dynamic strings for backwards compatibility sake so the `string` type
+// also implements the `Label` Kind
+c = "a" + ""
+fill(column: c, value: 0)
```

Comment:
I wonder how often a truly dynamic column name is used? Maybe we can get some of that from the query archive?


Also what do we consider dynamic? Anything that requires evalauting the semantic graph? What about this case:

```
c = "foo"
fill(column: c, value: 0)
```


Is that considered dynamic or static? Seems like it could be considered static. I would also expect this case to be very common as a user might have to use the column name in several different functions and so would like to put it in a variable to make changing it easier.

Reply:
Yes, that could be considered static in this context (probably harder to force it to be "dynamic" than not).